File: weighted_graph/dijkstra2/main.cpp
#include <iostream>
#include <vector>
#include <queue>
#define INFTY 1000000
#define UNVISITED 111
#define FOUND 222
#define VISITED 333

using namespace std;

int main(){
  int num_node;
  int k, u, v, c;

  cin >> num_node;
  vector<pair<int, int>> mat[num_node];
  vector<int> color, cost, parent;
  for(int i = 0; i < num_node; ++i){
    cin >> u >> k;
    for(int j = 0; j < k; ++j){
      cin >> v >> c;
      mat[u].emplace_back(make_pair(v, c));
    }
    color.push_back(UNVISITED);
    cost.push_back(INFTY);
    parent.push_back(-1);
  }

  priority_queue<pair<int /*cost*/, int/*id*/>> PQ;
  int start = 0;
  cost[start] = 0;
  color[start] = FOUND;
  PQ.push(make_pair(0, start));

  while(!PQ.empty()){
    auto f = PQ.top(); PQ.pop();
    int min_cost_open_node = f.second;
    color[min_cost_open_node] = VISITED;

    // 下のwhileでpushしたvはPQから削除される必要があるが，直接vをPQから削除できない
    // ここでcontinueされない場合，下のwhileで昔pushされた最短でないものを削除している
    if(cost[min_cost_open_node] < f.first * (-1)) continue;

    for(int j = 0; j < mat[min_cost_open_node].size(); ++j){
      int v = mat[min_cost_open_node][j].first;
      if(color[v] == VISITED) continue;
      int new_cost = cost[min_cost_open_node] + mat[min_cost_open_node][j].second;
      if(cost[v] > new_cost){
        cost[v] = new_cost;
        // priority_queue prioritize larger value
        PQ.push(make_pair(-cost[v], v));
        color[v] = FOUND;
        parent[v] = min_cost_open_node;
      }
    }
  }
  for(int i = 0; i < num_node; ++i)
    cout << i << ' ' << (cost[i] == INFTY ? -1 : cost[i]) << endl;
}

File: weighted_graph/dijkstra2/main.py
# -*- coding: utf-8 -*-
import sys
from queue import PriorityQueue

input = sys.stdin.readline

INFTY = float('inf')
UNVISITED = 111
FOUND = 222
FINISHED = 333

class Edge(object):
    def __init__(self, dst_vertex, edge_cost):
        self.dst_vertex = dst_vertex
        self.edge_cost = edge_cost

    @property
    def v(self):
        return self.dst_vertex

    @property
    def  c(self):
        return self.edge_cost
    
if __name__ == '__main__':
    num_node = int(input())
    adj_list = [[] for _ in range(num_node)]
    cost = [INFTY for _ in range(num_node)]
    color = [UNVISITED for _ in range(num_node)]
    parent = [-1 for _ in range(num_node)]

    for _ in range(num_node):
        inputs = list(input().split())
        u = int(inputs[0])
        k = int(inputs[1])
        for j in range(k):
            v = int(inputs[2*j+2])
            c = int(inputs[2*j+3])
            adj_list[u].append(Edge(v, c))

    queue = PriorityQueue()
    cost[0] = 0
    color[0] = FOUND
    queue.put((0, 0)) # in order (cost, vertex_id)
    while queue.qsize() is not 0:
        min_cost_found_node = queue.get()
        u = min_cost_found_node[1]
        c = min_cost_found_node[0]
        if cost[u] <  c:
            # this node is not treated
            continue

        color[u] = FINISHED
        for edge in adj_list[u]:
            v = edge.v
            if color[v] != FINISHED:
                # the cost of each node must be decrescent
                new_cost = c + edge.c
                if new_cost < cost[v]:
                    # cost lowers
                    cost[v] = new_cost
                    color[v] = FOUND
                    parent[v] = u
                    queue.put((cost[v], v))

    for i in range(num_node):
        print("{} {}".format(i, cost[i]))
